lib/ProductOpener/Auth.pm

```diff
@@ -0,0 +1,743 @@
+
+=head3 Return values
+
+Sign-in/sign-out cookie.
+
+=cut
+
+sub generate_oidc_cookie ($nonce, $return_url) {
+	my $signin_ref = {'nonce' => $nonce, 'return_url' => $return_url};
+
+	my $cookie_ref = {
+		'-name' => $cookie_name,
+		'-value' => $signin_ref,
+		'-path' => '/',
+		'-domain' => $cookie_domain,
+		'-samesite' => 'Lax',
+	};
+
+	return cookie(%$cookie_ref);
+}
+
+=head2 verify_access_token($access_token_string)
+
+Verifies the access token by decoding and validating it using the JSON Web Key Set (JWKS).
+
+Parameters:
+- $access_token_string: The access token to be verified.
+
+Returns: The verified access token or undefined if verification fails.
+=cut
+
+sub verify_access_token ($access_token_string) {
+	_ensure_oidc_is_discovered();
+
+	my $access_token_verified = decode_jwt(token => $access_token_string, kid_keys => $jwks);
+	$log->debug('access_token found', {access_token => $access_token_string, access_token => $access_token_verified})
+		if $log->is_debug();
+	unless ($access_token_verified) {
+		return;
+	}
+
+	return $access_token_verified;
+}
+
+=head2 verify_id_token($id_token_string)
+
+Verifies the ID token by decoding and validating it using the JWKS.
+
+Parameters:
+- $id_token_string: The ID token to be verified.
+
+Returns: The verified ID token or undefined if verification fails.
+=cut
+
+sub verify_id_token ($id_token_string) {
+	_ensure_oidc_is_discovered();
+
+	my $id_token = OIDC::Lite::Model::IDToken->load($id_token_string);
+	my $id_token_verified = decode_jwt(token => $id_token_string, kid_keys => $jwks);
+	$log->debug('id_token found', {id_token => $id_token, id_token_verified => $id_token_verified}) if $log->is_debug();
+	unless ($id_token_verified) {
+		return;
+	}
+
+	return $id_token_verified;
+}
+
+=head2 get_azp($access_token)
+
+Retrieves the authorized party (client ID) from the access token.
+
+=head3 Arguments
+
+=head4 The access token. $access_token
+
+=head3 Return values
+
+The authorized party (client ID) or undefined if the token is not issued by the correct issuer.
+=cut
+
+sub get_azp ($access_token) {
+	if (not(defined $access_token)) {
+		return;
+	}
+
+	_ensure_oidc_is_discovered();
+
+	if (not($oidc_discover_document->{issuer} eq $access_token->{iss})) {
+		$log->warn(
+			'Given token was not issued by the correct issuer',
+			{
+				actual_iss => $access_token->{iss},
+				expected_iss => $oidc_discover_document->{iss},
+				azp => $access_token->{azp},
+				sub => $access_token->{sub}
+			}
+		) if $log->is_warn();
+		return;
+	}
+
+	return $access_token->{azp};
+}
+
+sub _get_client () {
+	if ($client) {
+		return $client;
+	}
+
+	_ensure_oidc_is_discovered();
+	$client = OIDC::Lite::Client::WebServer->new(
+		id => $oidc_options{client_id},
+		secret => $oidc_options{client_secret},
+		authorize_uri => $oidc_discover_document->{authorization_endpoint},
+		access_token_uri => $oidc_discover_document->{token_endpoint},
+	);
+	return $client;
+}
+
+sub _ensure_oidc_is_discovered () {
```

Comment:
Can you comment to explain a bit the role of this ?

Reply:
Added a comment to the sub.